### ISSUE TITLE
Case for text/plain formatter to return unconditional toString()

### ISSUE THREAD

**Issue by chb0github** [restify/node-restify#952]
After quite a bit of research I discovered that 
'accept: text/plain' invokes the default handler (expected) but that handler tries to get too smart.

```
function formatText(req, res, body, cb) {
    if (body instanceof Error) {
        res.statusCode = body.statusCode || 500;
        body = body.message;
    } else if (typeof (body) === 'object') {
        body = JSON.stringify(body);
    } else {
        body = body.toString();
    }

    res.setHeader('Content-Length', Buffer.byteLength(body));
    return cb(null, body);
}
```

Now, if we look at the formatter for JSON:

```
function formatJSON(req, res, body, cb) {
    if (body instanceof Error) {
        // snoop for RestError or HttpError, but don't rely on
        // instanceof
        res.statusCode = body.statusCode || 500;

        if (body.body) {
            body = body.body;
        } else {
            body = {
                message: body.message
            };
        }
    } else if (Buffer.isBuffer(body)) {
        body = body.toString('base64');
    }

    var data = JSON.stringify(body);
    res.setHeader('Content-Length', Buffer.byteLength(data));

    return cb(null, data);
}
```

What we see is that, fundamentally, _text/plain_ and _application/json_ use the same code to produce JSON string data. Each is also inconsistent about handling errors.

I offer that each handler should call their respective functions:

```
function formatJSON(req, res, body, cb) {
   var data = body ? JSON.stringify(body) : {};
    res.setHeader('Content-Length', Buffer.byteLength(data));

    return cb(null, data);
}

function formatText(req, res, body, cb) {

    body = body ? body.toString() : '';
    res.setHeader('Content-Length', Buffer.byteLength(body));
    return cb(null, body);
}
```
Not only does this lead to a cleaner implementation but gives perfectly consistent (and polymorphic!) results to all objects passed in. This also means that object implementers can manipulate results by overriding toString();

I can make the change and submit a PR if it will actually be accepted

**Comment by DonutEspresso** [comment]
Thanks for logging the issue! What you've proposed here seems pretty reasonable to me. Thoughts, @yunong @micahr? 

Given that it'll be a breaking change, would be great to get this in for 5.x.

**Comment by micahr** [comment]
Seems reasonable to me as well. Look forward to the PR.

**Comment by yunong** [comment]
@chb0github That looks good, please send along a PR with tests.

**Comment by chb0github** [comment]
So, I implemented the changes but it exposed some problems.

Floating around in the code was stuff like this:
```
res.send(new InternalError(e, e.message || 'unexpected error'));
```
which introduces a cycle in the object because it magically scoops up execution context; I assume for debug purposes. But in reality, this could only ever have printed if the next line of code existed because it strips out this cycle (in the formatJSON() function):

```
body = body.body;
```
So, I removed the error-in-an-error situation everywhere it occurred in the tests and this removed the problem. However, then this code brings it's head to bear:

server.test.js:
```
test('returning error from a handler (with domains)', function (t) {
    SERVER.get('/', function (req, res, next) {
        next(new Error('bah!'));
    });

    CLIENT.get('/', function (err, _, res) {
        t.ok(err);
        t.equal(res.statusCode, 500);
        t.end();
    });
});

```
that _new Error('bah!')_ no long works because the formatJSON function no longer sets the error code; setting an error code was never a 'formatting' function and relied on the type checking.

This problem I don't know how to get past because I don't know the rest of the code well enough and I suspect what needs to happen is that another function in the chain can/should execute to determine what the error code should be (before formatting?) and set it; that would be good separation of concerns. 

This is the 1 test that I can't get to pass. I changed that next to:

```
        next(new errors.InternalError('bah!'));

```
And now a cycle is reintroduced because of the (in my opinion) way too clever method with which these errors are constructed. 

The best thing to do is implement separation of concerns. Alternatively, I can't change the JSON parser to be safe (and I have tested this) and remove the cycles; it is dangerous (as evidenced on the current problems) to ever assume a type for your data or manipulate the input on that assumption. To do that, I used 'json-stringify-safe' with an empty cycle function. This works but is not ideal. However, an ideal solution may be a significant amount of work. I just don't know.

I can checkin and push to a branch so you guys and see what I am saying and give me some guidance.

**Comment by DonutEspresso** [comment]
Thanks for digging into this! At first glance I'm not sure I follow the cycles issue. Please do push up a branch and I can take a look. From what I understand, the error objects created look something like:

```sh
$ node
> var e = require('restify-errors');
undefined
> var err = new e.InternalServerError('boom');
undefined
> err
{ [InternalServerError: boom]
  message: 'boom',
  body: { code: 'InternalServerError', message: 'boom' } }
> JSON.stringify(err)
'{"message":"boom","body":{"code":"InternalServerError","message":"boom"}}'
```

So the current JSON formatter, upon encountering an error, should just JSON.stringify the `body` property on the above error object. Agreed though, that the distributed type checking of error objects is less than ideal. 

I think we can try something like:

1) Remove type checking from the basic formatters (you've already done this)
2) Implementing both `toString()` and `toJSON()` into [restify-errors](https://github.com/restify/errors), thus handling both JSON.stringify and toString() called on error objects. Custom formatters will still need to handle errors via type checking, which seems reasonable to me. I can implement the change on this, and link back to this issue. 
3) Status codes for error objects should default to 500 here: https://github.com/restify/node-restify/blob/5.x/lib/response.js#L281

Does that cover all the missing items?

**Comment by chb0github** [comment]
I created a pull request [here](https://github.com/restify/node-restify/pull/988)

Forgive me if my PR is not in proper form. My first PR on GH ever.

**Comment by chb0github** [comment]
Hmm, How about something like this:

```
var data =  body.toString && body.toString();

var data = body.toJSON ? body.toJSON() : JSON.stringify(body);
```
This allows for the errors to polymorphically address their own idiosyncrasies and yet still allow every other object to get consistent behavior. I think I like this idea a lot.

**Comment by DonutEspresso** [comment]
Thanks! No worries about the PR, we're glad to get the help. :)

I dug into the issue, seems we've dived deep into the rabbit hole. There are a couple of things going on here:

1)The circular JSON is actually a result of domains tacking the domain object itself onto the error object, and of course JSON.stringify fails to stringify the domain, so it blows up. It  looks like we'd have to remove the domain references off the error object when stringifying - it may be something I'll add to restify/errors#18. @yunong lemme know if that sounds reasonable. I don't want to delete the domain objects. 

2) Response now needs to set a status code if the argument to `res.send()` is an error object. This was previously being done in formatters, but since you've removed it, it needs to be done here. 

3) Right where number two is happening, we should also wrap custom errors not generated by restify-errors using restify-errors module to ensure stable output. 

Lemme get restify-errors PR merged in and this should start to make more sense.

**Comment by chb0github** [comment]
1) That is what I realized too. I don't know what these "domain" objects are but they were deeply embedded so it seemed like it could be work to remove them or change them. 
2) How about there is a final handler that simply has something like:
```
var statusCode = errorObj.getStatusCode && errorObj.getStatusCode() || 500
```
This way, we never have to interrogate the object for what kind of error it is. We just ask what error code to set. This would require me knowing a lot more about the code base to effectively do that. 

3) I am not sure about wrapping. I mean, to get a consistent output would assume some kind of consistent input which you can never assume in a typeless language. If you try to, you will end up stuffing  anything that doesn't fit this predefined model into "everything else" fields

``` 
{
    statusCode: 500,
    originalObject: JSON.stringify(e)
}
That will just bring us right back to where we are. Error output should be free form. I was working on this at my last gig, but basically, I had a handler installed that grabbed all restify errors and gave them a consistent format, and then had a "if it's anything else, it's a 500". But it's best to add this as an optional handler with 
```
server.use(restify.consistentErrors()); 
```
or something like that. 

I think the toString() and toJSON() methods should be implemented in restify-errors/httpError.js for maxium consistency and maintainability sake. But because node-restify depends on this that's a bit out-of-scope for me. 

Bottom line: Based on my PR, how do you want me to proceed? Make the fixups that have been asked that I can or .. something else?

**Comment by DonutEspresso** [comment]
Sorry for dropping the ball on this! 

1) restify-errors has been updated, so the serialized errors should no longer contain the domain objects on them. They should support both toString() and JSON.stringify() serializations. 
2) Yeah, that sounds good - something along the lines of `if (e instanceof err) { res.send(e.statusCode || 500, err);`
3) I honestly don't remember the context for my comment on this anymore - if I remember I'll comment again. Probably ok to ignore this for now. 

TL;DR, the new serializations in restify-errors should allow you to proceed on 1 and 2 now!

**Comment by chb0github** [comment]
These should be seperation of concerns: 

1) convert to JSON whatever is passed in (error or otherwise, never inspect it)
2) Pass the object to an errorcode handler that then inspects the type or whatever to set the error code.

in number 2, that's another way of doing what I did. instead of checking the type, just check if there is a status code function. That way, You can map client exceptions (4xx) and server exceptions (5xx) to the same code space without having to check if it's your type or any type.

This is really a case for quack-typing: If the object is able to supply you with the error code then grab it. 
If it can't, then check and see if it's an error. If it's not an error, then it's a 200, ok?

**Comment by DonutEspresso** [comment]
Agree with #1, in this case any errors created by restify (and therefore restify-errors) will have built in serializers by default. This will strip the domain object and prevent the circular ref issue you were seeing earlier. I think we are good on this front. 

#2, I think checking for `e instanceof Error` is closer in spirit to the currently existing behavior. Presumably, users may want to be able to do something like this:

```js
res.json({
  data: { ... },
  prop1: '',
  prop2: '',
  statusCode: 409
});
```

And checking only for `statusCode` would cause the server to send a http 409 in this scenario, which is likely not what the user intended.

**Comment by DonutEspresso** [comment]
Hi @chb0github, thanks for all the time you've put into this so far! We really appreciate the help.

I'd like to get this PR merged in as we're looking to release 5.x imminently. With the fix to restify-errors to remove the serialization error, I think this PR is pretty much done. Do you think you'll have time to address @micahr's comments on the PR? If not, I'd be happy to branch it and clean it up. Let us know!

**Comment by chb0github** [comment]
Sorry to get back to you so belatedly. I didn't realize you guys were
expecting anything from me. I was more-or-less discussing approaches.

I will try to catch up on this discussion in code tonight.  I submitted the
PR a long while ago.

**Comment by DonutEspresso** [comment]
No worries! We hadn't planned on getting 5.x out until recently, as we decided reduce scope so we could get something out, since we've had a bunch of bug fixes that haven't yet been released, and we wanted to roll them up with the next major. Let me know where I can help.

**Comment by chb0github** [comment]
Ok guys, here is what I am going to do:

I am going to scrap that old PR and just put my changes into a new PR. It doesn't look like anything structurally has changed since my PR. 

Right now, I am getting linting errors around code that

1. I didn't touch
2. The error is actually wrong.

```
maximumLineLength: Line must be at most 80 characters at ./lib/formatters/text.js :
     1 |// Copyright 2012 Mark Cavage, Inc.  All rights reserved.
```

In my original change I removed this because it didn't seem to be in line with the license on the project; there was some discussion around doing that. So, I will leave it.

When I made the PR I was unemployed and had an itch to scratch. Now I am working and busy as hell. That means I don't have anytime to take this ball further. 

So, I will create a new PR based on the latest code with the understand that it's caveat emptor. When that's done I will update this ticket.

My overall thinking is that the chain of events around processing requests should be broken down into micro-steps for total separation of concerns. For example

- Formatting does just formatting and assumes no body structure at all.
- Setting the status code is entirely separate from anything else
- Setting headers should be 1 header (or logical set) per function.

**Comment by chb0github** [comment]
Ok gents,

See [this PR](https://github.com/restify/node-restify/pull/1074)

Sorry I can't do more

**Comment by DonutEspresso** [comment]
Appreciate it @chb0github! Thank you for all the contributions and the new PR! Good luck with the new gig.
